src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,462 @@
+}
+
+type CreatedPart = {
+    partId: string,
+    creator: MinimalUser,
+}
 
 describe('JoinerDAO', () => {
 
-    let dao: JoinerDAO;
+    let partDAO: PartDAO;
+    let joinerDAO: JoinerDAO;
+
+    const CREATOR_EMAIL: string = UserMocks.CREATOR_AUTH_USER.email.get();
+    const CREATOR_NAME: string = UserMocks.CREATOR_AUTH_USER.username.get();
+
+    const CANDIDATE_EMAIL: string = UserMocks.CANDIDATE_AUTH_USER.email.get();
+    const CANDIDATE_NAME: string = UserMocks.CANDIDATE_AUTH_USER.username.get();
+
+    const MALICIOUS_EMAIL: string = 'm@licio.us';
+    const MALICIOUS_NAME: string = 'malicious';
 
     beforeEach(async() => {
         await setupEmulators();
-        dao = TestBed.inject(JoinerDAO);
+        joinerDAO = TestBed.inject(JoinerDAO);
+        partDAO = TestBed.inject(PartDAO);
     });
     it('should be created', () => {
-        expect(dao).toBeTruthy();
+        expect(joinerDAO).toBeTruthy();
+    });
+
+    const defaultOptions: TestOptions = {
+        signOut: false,
+        createJoiner: false,
+    };
+    async function createPart(options: TestOptions = defaultOptions): Promise<CreatedPart> {
+        const creator: MinimalUser = await createConnectedUser(CREATOR_EMAIL, CREATOR_NAME);
+        const partId: string = await partDAO.create({ ...PartMocks.INITIAL, playerZero: creator });
+        if (options.createJoiner === true) {
+            const joiner: Joiner = { ...JoinerMocks.INITIAL, creator };
+            await joinerDAO.set(partId, joiner);
+        }
+        if (options.signOut === true) {
+            await signOut();
+        }
+        return { partId, creator };
+    }
+    async function addCandidate(partId: string, signOutUser: boolean = true): Promise<MinimalUser> {
+        const candidate: MinimalUser = await createConnectedUser(CANDIDATE_EMAIL, CANDIDATE_NAME);
+        await joinerDAO.addCandidate(partId, candidate);
+        if (signOutUser) {
+            await signOut();
+        }
+        return candidate;
+    }
+    describe('for non-verified users', () => {
+        it('should forbid to create a joiner', async() => {
+            // Given an existing part.
+            // (Note that the non-verified user in practice can never have created the corresponding part,
+            // but this is an extra check just in case)
+            const partId: string = (await createPart({ signOut: true })).partId;
+
+            // and a non-verified user
+            const nonVerifiedUser: MinimalUser = await createUnverifiedUser(MALICIOUS_EMAIL, MALICIOUS_NAME);
+
+            // When the user tries to create the joiner
+            const result: Promise<void> = joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator: nonVerifiedUser });
+
+            // Then it should fail
+            await expectPermissionToBeDenied(result);
+        });
+        it('should forbid to add themselves to the candidates', async() => {
+            // Given a part, and a non-verified user
```

Comment:
Tiens, y a plein d'endroit ou ce joiner/configRoom est appellé "part" alors que ce n'est pas ce document là qu'on teste.
Ca vaudrait le coup de rename ?

Reply:
il faut la part pour le joiner actuellement, donc je change en "given a part and joiner"